components/learn/modules/ROOT/pages/on-gsn.adoc

```diff
@@ -1,9 +1,338 @@
+
+[source,javascript]
+----
+import React, { useState, useEffect, useCallback } from "react";
+import { useWeb3Network } from "@openzeppelin/network/react";
+
+const PROVIDER_URL = "http://127.0.0.1:8545";
+
+function App() {
+  // get GSN web3
+  const context = useWeb3Network(PROVIDER_URL, {
+    gsn: { dev: true }
+  });
+
+  const { accounts, lib } = context;
+
+  // load Counter json artifact
+  const counterJSON = require("./build/contracts/Counter.json");
+
+  // load Counter Instance
+  const [counterInstance, setCounterInstance] = useState(undefined);
+
+  if (
+    !counterInstance &&
+    context &&
+    context.networkId
+  ) {
+    const deployedNetwork = counterJSON.networks[context.networkId.toString()];
+    const instance = new context.lib.eth.Contract(counterJSON.abi, deployedNetwork.address);
+    setCounterInstance(instance);
+  }
+
+  const [count, setCount] = useState(0);
+
+  const getCount = useCallback(async () => {
+    if (counterInstance) {
+      // Get the value from the contract to prove it worked.
+      const response = await counterInstance.methods.value().call();
+      // Update state with the result.
+      setCount(response);
+    }
+  }, [counterInstance]);
+
+  useEffect(() => {
+    getCount();
+  }, [counterInstance, getCount]);
+
+  const increase = async () => {
+    await counterInstance.methods.increase().send({ from: accounts[0] });
+    getCount();
+  };
+
+  return (
+    <div>
+      <h3> Counter counterInstance </h3>
+      {lib && !counterInstance && (
+        <React.Fragment>
+          <div>Contract Instance or network not loaded.</div>
+        </React.Fragment>
+      )}
+      {lib && counterInstance && (
+        <React.Fragment>
+          <div>
+            <div>Counter Value:</div>
+            <div>{count}</div>
+          </div>
+          <div>Counter Actions</div>
+            <button onClick={() => increase()} size="small">
+              Increase Counter by 1
+            </button>
+        </React.Fragment>
+      )}
+    </div>
+  );
+}
+
+export default App;
+
+----
+
+
+NOTE: You can pass a `dev: true` flag to the `gsn` options when setting up the provider. This will use the xref:gsn-provider::index.adoc[GSNDevProvider] instead of the regular GSN provider. This is a provider set up specifically for testing or development, and it _does not require a relayer to be running_ to work. This can make development easier, but it will feel less like the actual GSN experience. If you want to use an actual relayer, you can run `npx oz-gsn run-relayer` locally (see the xref:gsn-helpers::preparing-a-testing-environment.adoc[OpenZeppelin GSN helpers] for more info).
```

Comment:
```suggestion
NOTE: You can pass a `dev: true` flag to the `gsn` options when setting up the provider. This will use the xref:gsn-provider::index.adoc[GSNDevProvider] instead of the regular GSN provider. This is a provider set up specifically for testing or development, and it _does not require a relayer to be running_ to work. This can make development easier, but it will feel less like the actual GSN experience. If you want to use an actual relayer, you can run `npx oz-gsn run-relayer` locally (see the xref:gsn-helpers::preparing-a-testing-environment.adoc[Preparing a Testing Environment] for more info).
```